src/pad.rs

```diff
@@ -0,0 +1,363 @@
+        Ok(())
+    }
+
+    fn verify(&self, msg: untrusted::Input, encoded: untrusted::Input)
+              -> Result<(), ()> {
+        encoded.read_all((), |decoded| {
+            if try!(decoded.read_byte()) != 0 ||
+               try!(decoded.read_byte()) != 1 {
+                return Err(());
+            }
+
+            let mut ps_len = 0;
+            loop {
+                match try!(decoded.read_byte()) {
+                    0xff => { ps_len += 1; },
+                    0x00 => { break; },
+                    _ => { return Err(()); }
+                }
+            }
+            if ps_len < 8 {
+                return Err(());
+            }
+
+            let decoded_digestinfo_prefix =
+                try!(decoded.skip_and_get_input(
+                        self.digestinfo_prefix.len()));
+            if decoded_digestinfo_prefix != self.digestinfo_prefix {
+                return Err(());
+            }
+
+            let digest_alg = self.digest_alg;
+            let decoded_digest =
+                try!(decoded.skip_and_get_input(digest_alg.output_len));
+            let digest = digest::digest(digest_alg, msg.as_slice_less_safe());
+            if decoded_digest != digest.as_ref() {
+                return Err(());
+            }
+            Ok(())
+        })
+    }
+
+
+}
+
+// Maximum supported length of the salt;
+const MAX_SLEN: usize = 512 / 8;
+
+// Maxmimum supported length for intermediate value used in PSS padding.
+const MAX_MLEN: usize = 512;
+
+pub struct PSSPadding {
+    pub digest_alg: &'static digest::Algorithm,
+    pub mgf: fn(seed: &[u8], mask_len: usize, mask: &mut [u8],
+                digest_alg: &'static digest::Algorithm) -> Result<(), ()>,
+    pub s_len: usize
+}
+
+impl PaddingAlgorithm for PSSPadding {
+    // Implement padding procedure per EMSA-PSS,
+    // https://tools.ietf.org/html/rfc3447#section-9.1.
+    fn encode(&self, msg: &[u8], out: &mut [u8], rng: &rand::SecureRandom)
+              -> Result<(), ()> {
+
+        //    Steps:
+
+        //    1.  If the length of M is greater than the input limitation for the
+        //        hash function (2^61 - 1 octets for SHA-1), output "message too
+        //        long" and stop.
+
+        //    2.  Let mHash = Hash(M), an octet string of length hLen.
+
+        //    3.  If emLen < hLen + sLen + 2, output "encoding error" and stop.
+
+        let digest_len = self.digest_alg.output_len;
+        let em_len = out.len();
+        let min_len = digest_len + self.s_len + 2;
+
+        debug_assert!(em_len <= MAX_MLEN);
+        debug_assert!(self.s_len <= MAX_SLEN);
+
+        if em_len < min_len {
+            return Err(());
+        }
+        let m_hash = digest::digest(self.digest_alg, msg);
+
+        //    4.  Generate a random octet string salt of length sLen; if sLen = 0,
+        //        then salt is the empty string.
+
+        let mut salt = [0u8; MAX_SLEN];
+        try!(rng.fill(&mut salt[..self.s_len]));
```

Comment:
I would like us to try to find a deterministic method for calculating the salt, similar to [RFC 6979](https://tools.ietf.org/html/rfc6979). See the PSS spec which says:

> RSASSA-PSS is different from other RSA-based signature schemes in
   that it is probabilistic rather than deterministic, incorporating a
   randomly generated salt value.  The salt value enhances the security
   of the scheme by affording a "tighter" security proof than
   deterministic alternatives such as Full Domain Hashing (FDH); see [4]
   for discussion.  However, the randomness is not critical to security.
   In situations where random generation is not possible, a fixed value
   or a sequence number could be employed instead, with the resulting
   provable security similar to that of FDH [12].

I believe for ECDSA we will also be implementing the deterministic scheme.